ci: add action to retest failed approved PR's

Madhu-1 opened the pull request:
Adding Github action to retest the failed approved PR's. sample output is available at https://github.com/Madhu-1/retest-action/pull/3

Signed-off-by: Madhu Rajanna <madhupr007@gmail.com>

Madhu-1 committed:
ci: add action to retest failed approved PR's

Adding github action to retest the failed
approved PR's. sample output is available
at https://github.com/Madhu-1/retest-action/pull/3

Signed-off-by: Madhu Rajanna <madhupr007@gmail.com>

nixpanic commented:
Is it possible to maintain the action in a separate subdir, or branch of this repository?

nixpanic commented:
> [ ] Once schedule job runs and everything works as expected will send Doc PR

Better include that in this PR as as well. It helps with reviewing, we can then compare the intent and actual implementation.

Madhu-1 commented:
> Is it possible to maintain the action in a separate subdir, or branch of this repository?

action.yaml should be in the main directory haven't experimented with dockerfile path yet.

Madhu-1 commented:
> > [ ] Once schedule job runs and everything works as expected will send Doc PR
> 
> Better include that in this PR as as well. It helps with reviewing, we can then compare the intent and actual implementation.

my plan is to just cover asking the reviewers/maintainers to set the label to retest the PR. Are you looking for implementation details?

Madhu-1 commented:
Looks like having local actions are possible. let me add source code documentation etc so that we can have local action

nixpanic commented:
> > > [ ] Once schedule job runs and everything works as expected will send Doc PR
> > 
> > 
> > Better include that in this PR as as well. It helps with reviewing, we can then compare the intent and actual implementation.
> 
> my plan is to just cover asking the reviewers/maintainers to set the label to retest the PR. Are you looking for implementation details?

Something should be added to https://github.com/ceph/ceph-csi/blob/devel/docs/development-guide.md#retriggering-the-ci-jobs

Madhu-1 committed:
doc: add documentation for retest job

added details about retest job and the
creteria to auto retest PR.

Signed-off-by: Madhu Rajanna <madhupr007@gmail.com>

Madhu-1 committed:
ci: add retest github action

added source code of github retest
action.

Signed-off-by: Madhu Rajanna <madhupr007@gmail.com>

Madhu-1 committed:
ci: add github action to build retest

added basic github action for
retest building.

Signed-off-by: Madhu Rajanna <madhupr007@gmail.com>

Madhu-1 commented:
@nixpanic PTAL. moved the source code to cephcsi.

Madhu-1 committed:
ci: skip codespell for retest vendor

as there are lot of spell check failures
on the vendor directory, skipping it.

skipping retest action folder as
PullRequests word is not getting ignored

Signed-off-by: Madhu Rajanna <madhupr007@gmail.com>

Madhu-1 committed:
ci: skip shell check in vendor directory

skip shell check in vendor directories.

Signed-off-by: Madhu Rajanna <madhupr007@gmail.com>

humblec commented on actions/retest/README.md:
```diff
@@ -0,0 +1,15 @@
+# retest-action
+
+This is a github action built using the golang and the [github
+api](github.com/google/go-github). This main idea behind this one is to retest
```
`This main idea...` has to be refactored.

Madhu-1 replied:
open for suggestion

humblec commented on actions/retest/main.go:
```diff
@@ -0,0 +1,169 @@
+
+func main() {
+	if requiredlabel == "" {
+		log.Fatal("requiredlabels are not set")
```
`REQUIRED-LABEL` in the logs as other Vars?

Madhu-1 replied:
`required-label` should be set when running the action. `REQUIRED-LABEL` is the env internally created by github actions

nixpanic commented on scripts/codespell.conf:
```diff
@@ -1,4 +1,5 @@
 [codespell]
-skip = .git,./vendor,./docs/design/proposals/images
+# TODO: enable codespell on retest folder except vendor
+skip = .git,./vendor,./docs/design/proposals/images,./actions/retest/*
```
only add `./actions/retest/vendor` ?

Madhu-1 replied:
>codespell --config scripts/codespell.conf
./actions/retest/main.go:101: PullRequests ==> pull requests
./actions/retest/main.go:119: PullRequests ==> pull requests
./scripts/codespell.conf:4: PullRequests ==> pull requests
make: *** [Makefile:146: codespell] Error 65


codespell configuration is not working for the PullRequest word. so skipping retest folder for now

nixpanic replied:
Can `PullRequests` be added to the ignore words instead?

Madhu-1 replied:
see the `./scripts/codespell.conf:4: PullRequests ==> pull requests`  discard typo in above example. this is where we add the ignore list already tried it. its not working locally

nixpanic commented on actions/retest/Dockerfile:
```diff
@@ -0,0 +1,9 @@
+COPY . /home/src
+WORKDIR /home/src
+
+RUN go build -mod=vendor -o /bin/retest ./main.go
```
Building and using the container that contains all sources is not very clean. Consider using a multi-stage build and generating a container (`FROM scratch`) that contains only the `retest` executable.

Madhu-1 replied:
This is only testing locally when the action runs. not sure it makes sense to use multi-stage as we are not pushing image

nixpanic replied:
clean examples and best practices are important :smiley:

nixpanic commented on actions/retest/Dockerfile:
```diff
@@ -0,0 +1,9 @@
+WORKDIR /home/src
+
+RUN go build -mod=vendor -o /bin/retest ./main.go
+RUN chmod +x /bin/retest
```
this will be executable already, no need to chmod.

nixpanic commented on actions/retest/main.go:
```diff
@@ -0,0 +1,169 @@
+
+func main() {
+	if requiredlabel == "" {
+		log.Fatal("requiredlabel are not set")
```
This should this be `INPUT_REQUIRED-LABEL` or `required-label`, currently it seems inconsistent with the other errors.

Madhu-1 replied:
added as `required-label`. `INPUT_REQUIRED-LABEL` is not useful for user

nixpanic commented on actions/retest/main.go:
```diff
@@ -0,0 +1,169 @@
+// standard input env variables in action more details at
+// https://docs.github.com/en/actions/creating-actions/metadata-syntax-for-github-actions#inputs
+var (
+	retryLimit          = os.Getenv("INPUT_MAX-RETRY")
```
It would be cleaner to have a type `retestConfig`, fill that like you do now, and call a `retestConfig.validate()` function so that you can split validation of all the variables in a separate function.

nixpanic commented on actions/retest/main.go:
```diff
@@ -0,0 +1,169 @@
+		log.Fatalf("required-review-count %q is not valid", requiredReviewCount)
+	}
+
+	ctx := context.Background()
```
Only used once? just place it in `oauth2.NewClient()` directly. I also wonder why `Background()` and not `TODO()`.

nixpanic commented on actions/retest/main.go:
```diff
@@ -0,0 +1,169 @@
+		if *re.State == "open" {
+			prNumber := re.GetNumber()
+			log.Printf("PR with ID %d with Title %q is open\n", prNumber, re.GetTitle())
+			for _, r := range re.Labels {
```
Variable `r` used for `label`?

Madhu-1 replied:
it can be l. renaming it

nixpanic commented on actions/retest/main.go:
```diff
@@ -0,0 +1,169 @@
+					continue
+				}
+				log.Printf("checking status for PR %d with label %s", prNumber, r.GetName())
+				rs, _, err := client.Repositories.ListStatuses(context.TODO(), owner, repo, re.GetHead().GetSHA(), &github.ListOptions{})
```
move this to the for loop where it is used?

nixpanic commented on actions/retest/main.go:
```diff
@@ -0,0 +1,169 @@
+					log.Printf("failed to list comments %v\n", err)
+				}
+
+				for _, r := range rs {
```
re-using the `r` variable, this is ugly

Madhu-1 replied:
as using l variable fo label this should be fine.

nixpanic reviewed (changes requested):
Mostly comments related to the code maintenance of the added action. We should strive to have only high-quality code in the project, and that includes having comments and using helper functions where appropriate. The action is currently a single large function, that uses many variables and if-statements that can be split into smaller (unit-testable!) helpers.

humblec commented on actions/retest/README.md:
```diff
@@ -0,0 +1,15 @@
+* Check PR is open and have required approvals.
+* Check PR as the required label to continue to retest.
+* Pulls the failed test details.
+* Check if the retest of failed test is reached the maximum limit.
```
failed test has reached...

humblec commented on actions/retest/main.go:
```diff
@@ -0,0 +1,230 @@
+		}
+	}
+	if !(approvedReviews >= totalRequiredReviews) {
+		log.Printf("total approvred reviews for PR %d are %d but required %d", prNumber, approvedReviews, totalRequiredReviews)
```
approved.

nixpanic commented on actions/retest/Dockerfile:
```diff
@@ -0,0 +1,15 @@
+
+RUN go build -mod=vendor -o retest ./main.go
+
+FROM ${BASE_IMAGE}
```
use `FROM scratch` to only include the executable and none of the other golang stuff.

nixpanic commented on actions/retest/Dockerfile:
```diff
@@ -0,0 +1,15 @@
+
+COPY --from=builder ${WORK_DIR}/retest /usr/local/bin/retest
+
+ENTRYPOINT [ "usr/local/bin/retest" ]
```
should be `/usr/local/bin/retest`

nixpanic commented on actions/retest/Dockerfile:
```diff
@@ -0,0 +1,15 @@
+ARG WORK_DIR="/home/src"
+ARG BASE_IMAGE="golang:1.16"
+
+FROM ${BASE_IMAGE} as buidler
```
typo: `as builder`

nixpanic commented on actions/retest/main.go:
```diff
@@ -0,0 +1,230 @@
+	retry                int
+)
+
+type retestConfig struct {
```
thanks, this action is now much cleaner! :+1:

nixpanic reviewed (changes requested):
Just a few nits in the `Dockerfile` (might be renamed to `Containerfile` to suggest that other container runtimes can be used as well).

Great improvements since my previous review, thanks!

Madhu-1 commented:
> Just a few nits in the `Dockerfile` (might be renamed to `Containerfile` to suggest that other container runtimes can be used as well).
> 
keeping it as Dockerfile for now as it's used in most of the places we can do a cleanup to rename all dockerfiles in the repo.

> Great improvements since my previous review, thanks!

:+1:

nixpanic commented:
Verified that the new `<ci/retest/e2e>` label exists :smile:

humblec commented on actions/retest/README.md:
```diff
@@ -0,0 +1,15 @@
+* Check PR as the required label to continue to retest.
+* Pulls the failed test details.
+* Check failed test has reached the maximum limit.
+* If the limit is not reached the action will post the `retest` command on the
```
is/has

humblec commented on actions/retest/README.md:
```diff
@@ -0,0 +1,15 @@
+* Check failed test has reached the maximum limit.
+* If the limit is not reached the action will post the `retest` command on the
+  PR with log location for further debugging.
+* If the limit is reached the Pull Request will be skipped.
```
is/has

humblec commented on actions/retest/main.go:
```diff
@@ -0,0 +1,230 @@
+	return c
+}
+
+func (c retestConfig) validate() error {
```
would be good to add some comment about the functions here and below

humblec commented on actions/retest/main.go:
```diff
@@ -0,0 +1,230 @@
+							continue
+						}
+						if ok {
+							log.Printf("Pull Requested %d: %q reached  maximum attempt. skipping retest %v\n", prNumber, r.GetContext(), retry)
```
Requested/Request

Rakshith-R commented:
@mergifyio rebase
